docs/root/adaptive_load_controller.md

```diff
@@ -0,0 +1,202 @@
+# The Adaptive Load Controller
+
+The adaptive load controller is a standalone
+[library](https://github.com/envoyproxy/nighthawk/tree/main/source/adaptive_load)
+with [proto based
+API](https://github.com/envoyproxy/nighthawk/tree/main/api/adaptive_load). It is
+an abstraction built on to of Nighthawk that implements an adaptive execution mode.
+
+## Motivation
+
+When executing Nighthawk in [closed-loop](terminology.md#closed-loop) or
+[open-loop](terminology.md#open-loop) mode, the test configuration must specify
+the load level to deliver to the system under test, i.e. the number of requests
+per second (RPS). When running tests continuously over a longer time period,
+this configuration needs to be manually maintained and may get obsolete as the
+performance of the system under test changes over time.
+
+When executing Nighthawk in the adaptive mode, the adaptive load controller
+automatically determines the maximum load specification the system under test
+can sustain and still meet some defined criteria (e.g. success rate, latency,
+etc...). This decreases test maintenance and allows to monitor how the
+performance of the system under test evolves over time. The maximum load
+specification is further referred to as the optimal configuration.
+
+## The adaptive mode
+
+The execution is broken into two stages in the adaptive mode:
+
+*  The adaptive (search) stage.
+*  The testing stage.
+
+![Adaptive Mode](images/adaptive_mode.png)
+
+In the **adaptive stage** the adaptive load controller iteratively executes
+Nighthawk with varying test parameters in search for the optimal configuration.
+The individual Nighthawk executions use the
+[open-loop](terminology.md#open-loop) mode. The controller collects metrics
+after each iteration and evaluates them according to configured thresholds in
+order to determine of the configuration used in the iteration is optimal. Once
+the controller finds the optimal test configuration, it enters the testing
+stage.
+
+In the **testing stage** the adaptive load controller delivers load to the
+system under test using the optimal test configuration found in the adaptive
+stage.
+
+### Example adaptive mode execution
+
+The following diagram illustrates an execution when using the adaptive mode.
+
+![Example Adaptive Mode Execution](images/example_adaptive_mode_execution.png)
+
+## Architecture
+
+This diagram outlines the architecture of the adaptive load controller:
+
+![Adaptive Controller Architecture](images/adaptive_controller_architecture.png)
+
+The **adaptive load controller** binary talk to a Nighthawk server over gRPC.
```

Comment:
talks

Reply:
Done in e70312eba918d9646b65e8f7f92b20b59437d94e.